fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/MessageValue.java

```diff
@@ -105,6 +105,23 @@ private static Message getFieldMessageOnMessage(@Nonnull MessageOrBuilder messag
         return null;
     }
 
+    public static <M extends Message> Object getNullableField(Message message, int fieldNumber) {
+        final Descriptors.FieldDescriptor field = findFieldDescriptorOnMessageByNumber(message, fieldNumber);
+        if (field.isRepeated()) {
+            int count = message.getRepeatedFieldCount(field);
+            List<Object> list = new ArrayList<>(count);
+            for (int i = 0; i < count; i++) {
+                list.add(message.getRepeatedField(field, i));
+            }
+            return list;
+        }
+        if (field.hasDefaultValue() || message.hasField(field)) {
+            return message.getField(field);
+        } else {
+            return null;
+        }
+    }
+
```

Comment:
This method looks very similar to `Object getFieldOnMessage(@Nonnull MessageOrBuilder message, @Nonnull String fieldName)`, except that the later one takes care of special `TupleFields` which is good I guess to preserve backward compatibility.

Therefore, I would suggest to reuse `getFieldOnMessage ` and refactor it accordingly such that we have an overload of it that returns the requested field value by number, which is what you want as far as I can tell, i.e.:

```java
    @Nullable
    public static Object getFieldOnMessage(@Nonnull MessageOrBuilder message, @Nonnull String fieldName) {
        final Descriptors.FieldDescriptor field = findFieldDescriptorOnMessage(message, fieldName);
        return getFieldOnMessage(message, field);
    }

    @Nullable
    public static Object getFieldOnMessage(@Nonnull MessageOrBuilder message, int fieldNumber) {
        final Descriptors.FieldDescriptor field = findFieldDescriptorOnMessageByNumber(message, fieldNumber);
        return getFieldOnMessage(message, field);
    } 
    
    @Nullable
    private static Object getFieldOnMessage(@Nonnull MessageOrBuilder message, @Nonnull Descriptors.FieldDescriptor field) {
        if (field.isRepeated()) {
            int count = message.getRepeatedFieldCount(field);
            List<Object> list = new ArrayList<>(count);
            for (int i = 0; i < count; i++) {
                list.add(message.getRepeatedField(field, i));
            }
            return list;
        }
        if (field.hasDefaultValue() || message.hasField(field)) {
            if (field.getJavaType() == Descriptors.FieldDescriptor.JavaType.MESSAGE &&
                     TupleFieldsHelper.isTupleField(field.getMessageType())) {
                return TupleFieldsHelper.fromProto((Message)message.getField(field), field.getMessageType());
            } else {
                return message.getField(field);
            }
        } else {
            return null;
        }
    }
```

Reply:
👍